### ISSUE TITLE
Undo recipe

### ISSUE THREAD

**Issue by HansDP** [yelouafi/redux-saga#647]
Hello

I was looking at the Undo recipe shown on the documentation website of redux-saga. 

I love the idea, it is plain and simple. However, I started wondering whether there are issues with the provided code.

This is what got me thinking (so, please correct me if I'm wrong, which would not surprise me, since I'm not an expert, yet :)

When the main saga starts, a `takeEvery` (which is non-blocking) will listen for all `ARCHIVE_THREAD` actions and immediately continue to the next `take(UNDO)` action, which is blocking. Now, when the user performs an ’archive thread’, an `onArchive` is forked. When the user performs an `UNDO`, the `take(UNDO)` gets unblocked, the next statement will cancel all forks created by the `takeEvery` task and the `while(true)` gets restarted. All is good. 

The same applies when the user does not `UNDO` the archive (this will just keep the `takeEvery` waiting for the next archive action). 

But, what happens when the user triggers two ’archive threads’ in rapid succession? Probably two ’undo UI elements’ will be shown. But when the user now does an `UNDO` on any of the ’undo UI elements’, both archive requests will be undone. Or, maybe, the UI is smart enough to combine both UI elements into one (this depends on the requirements of the application). Then it would make more sense. 

Another issue I see with this solution is when the user triggers an `UNDO` at the exact time where the code is calling the `updateThreadApi` method. This method is probably a Fetch request (or similar). 
So, the HTTP request has been started, and now the `archiveTask` gets cancelled, forcing the `onArchive` to execute the `finally` block and thus marking the thread as not archived. And this results in an invalid state (the thread is archived on the server, but the client thinks it is not archived)
 
All of this made me wonder why this Undo recipe was not provided using a race condition? In my personal opinion, using a race would make the code much more understandable. So, I tried to make a solution (small disclaimer: I did not test this code, I've written it in Notes on my phone :-s

``` javascript 

function* onArchive(action) {
  const { threadId } = action
  const undoId = `UNDO_ARCHIVE_${threadId}`
  
  const thread = { id: threadId,archived: true }
  // show undo UI element, and provide a key to communicate
  yield put(actions.showUndo(undoId))
  // optimistically mark the thread as `archived`
  yield put(actions.updateThread(thread))
  // allow user time to active the undo action
  const { undo, timedout } = yield race({
    undo: take(action => action.type === UNDO && action.undoId === undoId),
    timedout: call(delay, 5000)
  })
      
  // hide undo UI element, in any case
  yield put(actions.hideUndo(undoId))
      
  if (undo) {
    // revert thread to previous state
    yield put(actions.updateThread({ id: threadId, archived: false }))
  } else if (timedout) {
    // make the API call to apply the changes remotely
    yield call(updateThreadApi, thread)
  }
}

function* main() {
  while (true) {
  
    const action = take(`ARCHIVE_THREAD`)
    // use spawn to execute non-blocking and prevent cancelation when main saga gets cancelled. this is needed to keep all state in sync (server and client)
    spawn(onArchive, action)
  }
}
```

What do you think?

Regards
Hans

**Comment by Andarist** [comment]
You are right that this recipe is not ideal and your snippet fixes its issues.

Would you maybe like to make pull request out of it?
